### ISSUE TITLE
Dart imports are not case sensitive

### ISSUE THREAD

**Issue by AbhishekDoshi26** [dart-lang/sdk#58663]
## Summary of Issue:
Recently encountered an issue which depicts that flutter's imports are not case sensitive. So for Flutter Engine:
`import 'package:test_project/screens/home.dart';` and `'package:test_project/Screens/home.dart';` are same which is totally incorrect. Now if the user imports as 2nd option, other services like firebase Hosting, Docker, GitHub actions, fail as they all are case sensitive and for them, this import will be undefined.

In one sentence, my folder name is `screen` and I imported it with `Screen` and still it works without any issues!

## Steps to Reproduce

1. Create a new project with main.dart as the basic file
2. Create another folder called screens and inside that create a file called home.dart
3. Import the file in main.dart and change the import to have it `Screen\home.dart` instead of `screen\home.dart`

**Expected results:** Dart Analyzer should throw an error for import not found as `Screen` and `screen` should be 2 different folders.

**Actual results:** It works! No errors, no warnings!

**Found on:** macOS 12.2.1, Flutter version 2.8.1 as well as 2.10.2, Dart version 2.15.1 as well as 2.16.1
<details>
<summary>Code sample</summary>

#### main.dart
```
import 'package:flutter/material.dart';
import 'package:test_project/Screens/home.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      title: 'Material App',
      home: Home(),
    );
  }
}


```

#### home.dart
```
import 'package:flutter/material.dart';

class Home extends StatelessWidget {
  const Home({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Container();
  }
}


```
</details>

<details>
  <summary>Logs</summary>

#### flutter run
```
Launching lib/main.dart on Chrome in debug mode...
Waiting for connection from debug service on Chrome...             12.5s
This app is linked to the debug service: ws://127.0.0.1:60085/QgzYCoQNVzo=/ws
Debug service listening on ws://127.0.0.1:60085/QgzYCoQNVzo=/ws

💪 Running with sound null safety 💪

🔥  To hot restart changes while running, press "r" or "R".
For a more detailed help message, press "h". To quit, press "q".

An Observatory debugger and profiler on Chrome is available at: http://127.0.0.1:60085/QgzYCoQNVzo=
The Flutter DevTools debugger and profiler on Chrome is available at: http://127.0.0.1:9104?uri=http://127.0.0.1:60085/QgzYCoQNVzo=

```

#### flutter analyze

```
Analyzing test_project...                                               
No issues found! (ran in 1.5s)
```

#### flutter doctor -v

```
[✓] Flutter (Channel stable, 2.8.1, on macOS 12.2.1 21D62 darwin-arm, locale en-GB)
    • Flutter version 2.8.1 at /Users/abhishekdoshi/Desktop/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 77d935af4d (10 weeks ago), 2021-12-16 08:37:33 -0800
    • Engine revision 890a5fca2e
    • Dart version 2.15.1

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/abhishekdoshi/Library/Android/sdk
    • Platform android-31, build-tools 31.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2020.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)

[✓] VS Code (version 1.64.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.34.0

[✓] Connected device (1 available)
    • Chrome (web) • chrome • web-javascript • Google Chrome 98.0.4758.102

• No issues found!
```

</details>

**Comment by FMorschel** [comment]
Duplicate of https://github.com/dart-lang/sdk/issues/17064 but this has more discussion so I guess the older issue can be closed.

**Comment by stereotype441** [comment]
@cj-radcliff recently ran into this issue (specifically the sort of scenario described in https://github.com/dart-lang/sdk/issues/58663#issuecomment-2486536729). In @cj-radcliff's case the problem was that the program was interpreted by the compiler as having two different copies of the same file (differing only in the case of their filenames) and so the types defined in that file were duplicated, and so a runtime type check was failing that was expected to succeed.

Here's a minimal repro:
```dart
// common.dart
class C {}

// producer.dart
import 'common.dart';

Object produce() {
  return C();
}

// consumer.dart
import 'Common.dart'; // Oops, should be `common.dart`

void consume(Object o) {
  if (o is! C) {
    print('Got an object of type ${o.runtimeType}, expected type C');
  }
}

// main.dart
import 'producer.dart';
import 'consumer.dart';

main() {
  consume(produce());
}
```

If I run this example on my computer (which has a case sensitive filesystem), I get a nice helpful compile-time error:
```
consumer.dart:1:8: Error: Error when reading 'Common.dart': No such file or directory
import 'Common.dart'; // Oops, should be `common.dart`
       ^
consumer.dart:4:13: Error: 'C' isn't a type.
  if (o is! C) {
            ^
```

But I believe users with a case <i>in</i>sensitive filesystem will see this very confusing output:
```
Got an object of type C, expected type C
```

**Comment by lrhn** [comment]
@stereotype441 Yep. \<meme>It used to be a problem. It still is, but it used to too. \</meme>

It's a _hard to fix_ problem.
Even the draconic solution of saying that a Dart file, or any Dart package internal directory, must not contain any upper-cased letters can still run into issues when the path to the package directory does. Like `/Users/banana/dart/pkgs/foo`.
If one place in the code uses `file:///users/...` and another `file:///Users/...` as base URI, they'll still disagree on the URI of the file.

The only way to check if two URIs point to the same file system file is to ask the file system.
We can't assume that Windows is case-insensitive. It's the default, but it can be changed on directory-to-directory basis.
We can't assume that MacOS is case-sensitive or case-insensitive. It can be changed, probably per mount. People seem to actually do that.
Can't even assume Linux is case-insensitive, case insensitivity is built into the kernel.

So we have to ask the file system if a file exists. We can *try* to ask the file system if two entries are the same one, but that's non-trivial. If a case sensitive file system contains both `foo.txt` and `Foo.txt` *hard-linked to the same inode*, how do we determine whether they are "the same file"? 
Can we list the directory and check if we find both? But how do we even *compare names*? We can't just use Dart's `.toUpperCase()`, we don't know that's what the file system uses. Windows has a per-volume case-folding table, and it has changed over time. The case folding may depend on when the drive was first formatted! So we'd need to read that file and understand it, just to determine if the `FileSystemEntity.name` we get from `Directory.list` matches the name in our path.

File systems are many-to-one mappings from paths to file system entities, "I have a path, give me an object" with multiple paths being able to refer to the same object for many different reasons.

File systems won't just tell you whether _path1_ and _path2_ point to the same entity. Either you create your own paths and use and keep track of those, or you ask the file system for which entries exist (by listing directories) and then use the name given by that.

Determining whether two *similar* paths denote the same entity, without making two *actually intended to be different* paths be equal, just because they happen to denote the same inode, means figuring out whether the paths have the same length, and each segment denotes the same *entry of its directory*.

There is no good way to do that. You can't look up the entry of a directory by name, you have to list them all and then check if the name matches, and checking if two different names are the same to the file system is precisely the problem we can't solve to begin with.

The `dart:io` `FileSystem.identicalSync` function looks promising, but it actually checks the inode. If we have a hard-link of the same file with two different names, it would say they are the same. (It probably works for directories when there are no hard-links for directories. I don't know whether a Windows directory junction should count as a soft or hard link.)

Otherwise something like this would have worked:
```dart
import "package:path/path.dart" as p;

/// Whether two *normalized* paths denote the same path through the same directories.
///
/// Must not contain any `.` or `..` segments. If relative, should be relative to the
/// same path.
///
/// Only needed for a case-insensitive file system where multiple path segment
/// strings can denote the same directory entry.
///
/// Checks if the two paths exist and `FileSystemEntity.identicalSync` considers
/// them identical, as well as every path segment in the path also being the
/// same directory.
bool sameFileSystemEntity(String path1, String path2) {
  // Guard `identicalSync` against the entry not existing.
  var type1 = FileSystemEntity.typeSync(path1);
  var type2 = FileSystemEntity.typeSync(path2);
  if (type1 != type2) return false; // Definitely not the same.
  if (type1 == FileSystemEntityType.notFound) return false;
  // Check if paths point to same entry through the same path.
  while (path1 != path2) { // Should happen no later than when we hit the root.
    if (!FileSystemEntity.identicalSync(path1, path2, followLinks: false)) return false;
    path1 = p.dirname(path1);
    path2 = p.dirname(path2);
  }
  return true;
}
```

TL;DR: To solve the question of "are these two different paths the same after case canonicalization", we either need to reimplement the actual case canonicalization performed by the file system, which can change per volume on Windows, and with junction points you can have multiple volumes in one path, or we need a way to identify or distinguish the directory entries of two names in a directory. The `dart:io` file system does not provide such a feature (AFAICS).

**Comment by techouse** [comment]
I just encountered this bug in the wild.

My file is `foo_bar.dart`
```dart
// foo_bar.dart
final String baz = 'qux';
```

and I imported like this in my main.dart
```dart
// main.dart
import 'foo_Bar.dart`

print(baz);
```

The odd thing is, that the app compiles (on macOS), but if I reference anything from `foo_bar.dart`, i.e. `baz`, I get a runtime error.

I would have expected this to be a compile error.

**Comment by lrhn** [comment]
A runtime error does sound weird. Which runtime error?

I assume the `main.dart` should have a `main` function around the print. Can you show the exact files that cause the error, say how you run them and what the error is?
